src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/DefaultRazorLanguageServerCustomMessageTarget.cs

```diff
@@ -296,11 +299,62 @@ public override async Task<ProvideSemanticTokensResponse> ProvideSemanticTokensA
                 semanticTokensParams,
                 cancellationToken).ConfigureAwait(false);
 
-            var result = new ProvideSemanticTokensResponse(csharpResults, csharpDoc.HostDocumentSyncVersion);
+            var result = new ProvideSemanticTokensResponse(csharpResults, editResult: null, csharpDoc.HostDocumentSyncVersion);
 
             return result;
         }
 
+        [Obsolete]
+#pragma warning disable CS0809 // Obsolete member overrides non-obsolete member
+        public override async Task<ProvideSemanticTokensResponse> ProvideSemanticTokensEditsAsync(
+#pragma warning restore CS0809 // Obsolete member overrides non-obsolete member
+            SemanticTokensEditsParams semanticTokensEditsParams,
+            CancellationToken cancellationToken)
+        {
+            if (semanticTokensEditsParams is null)
+            {
+                throw new ArgumentNullException(nameof(semanticTokensEditsParams));
+            }
+
+            if (!_documentManager.TryGetDocument(semanticTokensEditsParams.TextDocument.Uri, out var documentSnapshot))
+            {
+                return null;
+            }
+
+            if (!documentSnapshot.TryGetVirtualDocument<CSharpVirtualDocumentSnapshot>(out var csharpDoc))
+            {
+                return null;
+            }
+
+            semanticTokensEditsParams.TextDocument.Uri = csharpDoc.Uri;
+
+            var csharpResults = await _requestInvoker.ReinvokeRequestOnServerAsync<SemanticTokensEditsParams, SumType<LanguageServer.Protocol.SemanticTokens, SemanticTokensEdits>>(
+                LanguageServerConstants.LegacyRazorSemanticTokensEditEndpoint,
+                LanguageServerKind.CSharp.ToContentType(),
+                semanticTokensEditsParams,
+                cancellationToken).ConfigureAwait(false);
+
+            // Converting from LSP to O# types
+            if (csharpResults.Value is LanguageServer.Protocol.SemanticTokens tokens)
+            {
+                var convertedTokens = new SemanticTokens { ResultId = tokens.ResultId, Data = ImmutableArray.Create(tokens.Data) };
```

Comment:
Hmmm, we may want to create a client side version of ProvideSemanticTokensResponse so we don't have to do `ImmutableArray.Create`, I imagine that can get quite expensive when all we're really doing here is serializing the content